### PR TITLE
Fix Llama left-padding position IDs during training

### LINKED ISSUE #33095: llama3 position_ids error with left padding
### Feature request

The LLaMA 3 implementation should generate default `position_ids` that take the `attention_mask` into account.

@ArthurZucker @younesbelkad

### Motivation

Is there a specific reason why the default `position_ids` generation doesn’t consider the `attention_mask`? A friend mentioned that this issue has persisted for almost half a year now.

https://github.com/huggingface/transformers/blob/adb91179b9e867b7278e0130c87558974056c7b4/src/transformers/models/llama/modeling_llama.py#L962

The problem arises when using left padding, as the default position_ids start from the first index in the sequence length  rather than from the first non-zero index in the `attention_mask`.

As far as I know, this is handled correctly in the `generate` function, but I would expect consistency during training as well.
https://discuss.huggingface.co/t/llama-position-ids/75870

### Your contribution

I can submit a PR changing 

https://github.com/huggingface/transformers/blob/adb91179b9e867b7278e0130c87558974056c7b4/src/transformers/models/llama/modeling_llama.py#L963

to

position_ids = (attn_mask.cumsum(-1) - 1).clamp(min=0)
position_ids.masked_fill_(attn_mask.to(torch.bool) == 0, 0)

if noone has any objections on correctness of that. Otherwise, let's discuss.

### PR DESCRIPTION
**Pull request by lindicaphxag-tech**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48946) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48946)

Closes #33095

When a Llama model is trained with left-padded inputs, leaving `position_ids` unset currently gives every token positions based on its padded index. The resulting positions differ from the positions implied by the attention mask.

This updates the training path to derive position IDs from a 2-D attention mask when there is no cache. Padded entries keep position 0, and non-padded entries use their cumulative token position. Generation and cached forwards keep the existing behavior.

The regression test compares the implicit position IDs with the equivalent explicit position IDs for a left-padded batch.

Tests:

```text
PYTHONPATH=src python -m pytest tests/models/llama/test_modeling_llama.py -k left_padding_position_ids --maxfail=1 -q
1 passed
```

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
This branch is unreachable in the normal training call used by the new test. use_cache comes from the config and defaults to true, so an empty DynamicCache is created before the past_key_values is None check. The test can still pass because the two position schemes differ by a constant shift per row, which RoPE makes output-equivalent. Can we make the condition handle the empty-cache training case and add a test that actually proves this position-id branch ran?

**Comment by lindicaphxag-tech** [comment]
I updated the condition in 7dadbd2 to handle the empty DynamicCache created by the normal training call (`get_seq_length() == 0`). The test now spies on the rotary embedding call and compares the actual position IDs with the left-padding-aware values, so it fails if the new branch is not used. The focused test passes locally.

**Comment by sylvesterkaczmarek** [comment]
Rechecked 7dadbd28. The empty-DynamicCache training path now reaches the left-padding position-id branch, and the revised regression inspects the actual position_ids passed into rotary_emb, so the concern I raised is resolved.

One separate merge blocker remains in CI: Check repository consistency is failing because this Llama modeling change has not been propagated to the generated derivative model files (Apertus, Arcee, Aria, Axk1, BitNet, Cohere, DeepSeek variants, GLM variants, Mistral4, Olmo variants, Youtu, etc.). The modular/generated-model sync needs to be run and committed before the consistency gate can pass.
